### ISSUE TITLE
MoltenVK 1.0: Depth/stencil attachment as input and output attachment in the same subpass

### ISSUE THREAD

**Issue by aitor-lunarg** [KhronosGroup/Vulkan-Portability#66]
Failing CTS tests:
```
dEQP-VK.rasterization.rasterization_order_attachment_access.depth.samples_1.multi_draw_barriers
dEQP-VK.rasterization.rasterization_order_attachment_access.stencil.samples_1.multi_draw_barriers
dEQP-VK.rasterization.rasterization_order_attachment_access.stencil.samples_2.multi_draw_barriers
```

Reasoning:
Metal does the depth/stencil write before the fragment execution, modifying the input attachment and generating a different output from the expected in the test. However, it is not clear to me if these tests are Vulkan conformat. Created the following issue to clarify: https://gitlab.khronos.org/Tracker/vk-gl-cts/-/issues/5087

Solution:
Wait for resolution of mentioned issue. If the conclusion is that the tests are valid, we can force the depth/stencil write in Metal to post fragment shader execution if we write the depth value in the shader as pass through.

**Comment by aitor-lunarg** [comment]
Conclusion of https://gitlab.khronos.org/Tracker/vk-gl-cts/-/issues/5087 is that the tests are valid.

In short, Vulkan expects depth write to happen at `VK_PIPELINE_STAGE_2_LATE_FRAGMENT_TESTS_BIT` but Metal may optimize the write in `VK_PIPELINE_STAGE_2_EARLY_FRAGMENT_TESTS_BIT` before execution of `VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT` which is not valid Vulkan.

Closing in favor of https://github.com/KhronosGroup/MoltenVK/issues/2232 since we can circumvent the issue in MoltenVK.

**Comment by theFJ1991** [comment]
F J XROn 8 May 2024, at 19:50, Aitor Camacho ***@***.***> wrote:﻿
Closed #66 as completed.
